File: flutter/lib/api/internal/event/dwell_tracker.dart
/// Measures how long an experience was on screen — the `dwell_ms` recorded on
/// dismiss/complete analytics events.
///
/// [markViewed] stamps the view time keyed by `cepCampaignId`; [consumeDwellMs]
/// reads the elapsed time and forgets the mark in one step (the typical dismiss
/// path), while [elapsedMs] peeks without forgetting.
///
/// Mirrors Android's `DwellTracker`.
class DwellTracker {
  final int Function() _now;

  /// [now] is injectable for tests; defaults to wall-clock milliseconds.
  DwellTracker({int Function()? now})
      : _now = now ?? (() => DateTime.now().millisecondsSinceEpoch);

  final Map<String, int> _viewedAtMs = {};

  /// Stamps the moment [cepCampaignId] became visible.
  void markViewed(String cepCampaignId) => _viewedAtMs[cepCampaignId] = _now();

  /// Milliseconds since [cepCampaignId] was viewed, or null if never marked.
  int? elapsedMs(String cepCampaignId) {
    final viewedAt = _viewedAtMs[cepCampaignId];
    return viewedAt == null ? null : _now() - viewedAt;
  }

  /// Like [elapsedMs] but also forgets the mark, so the next showing starts
  /// fresh. Returns null if [cepCampaignId] was never marked.
  int? consumeDwellMs(String cepCampaignId) {
    final viewedAt = _viewedAtMs.remove(cepCampaignId);
    return viewedAt == null ? null : _now() - viewedAt;
  }

  /// Forgets every view mark.
  void clear() => _viewedAtMs.clear();
}

File: android/digia-engage/src/main/java/com/digia/engage/internal/event/DwellTracker.kt
package com.digia.engage.internal.event

/**
 * Tracks how long each campaign instance was on screen.
 *
 * [markViewed] stamps the moment a campaign becomes visible; [consumeDwellMs]
 * returns the elapsed time at dismissal and forgets the mark. Keyed by
 * `cepCampaignId`. SRP: this owns dwell timing only, decoupled from the events
 * and the emitter. The clock is injectable for tests.
 */
internal class DwellTracker(
    private val now: () -> Long = { System.currentTimeMillis() },
) {
    private val viewedAtMs = mutableMapOf<String, Long>()

    /** Records that [cepCampaignId] became visible now. */
    fun markViewed(cepCampaignId: String) {
        viewedAtMs[cepCampaignId] = now()
    }

    /**
     * Returns ms since [markViewed] for [cepCampaignId] *without* forgetting the
     * mark — for mid-life signals (e.g. a click while the campaign is still up).
     */
    fun elapsedMs(cepCampaignId: String): Long? =
        viewedAtMs[cepCampaignId]?.let { now() - it }

    /**
     * Returns ms elapsed since [markViewed] for [cepCampaignId] and forgets the
     * mark, or null if it was never marked (so callers omit the field).
     */
    fun consumeDwellMs(cepCampaignId: String): Long? =
        viewedAtMs.remove(cepCampaignId)?.let { now() - it }

    fun clear() {
        viewedAtMs.clear()
    }
}
